sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala

```diff
@@ -297,50 +329,142 @@ class KudoTableOperator(kudo: Option[KudoSerializer], readOption: CoalesceReadOp
   }
 }
 
+
+class KudoGpuTableOperator(dataTypes: Array[DataType])
+  extends SerializedTableOperator[KudoSerializedTableColumn, ColumnarBatch] {
+
+  override def getDataLen(column: KudoSerializedTableColumn): Long =
+    column.spillableKudoTable.header
+      .getTotalDataLen
+
+  override def getNumRows(column: KudoSerializedTableColumn): Int =
+    column.spillableKudoTable.header
+      .getNumRows
+
+  override def concat(columns: Array[KudoSerializedTableColumn]): ColumnarBatch = {
+    require(columns.nonEmpty, "no tables to be concatenated")
+    val numCols = columns.head.spillableKudoTable.header.getNumColumns
+    if (numCols == 0) {
+      val totalRowsNum = columns.map(getNumRows).sum
+      new ColumnarBatch(Array.empty, totalRowsNum)
+    } else {
+      withResource(columns.safeMap(_.spillableKudoTable.makeKudoTable)) { kudoTables =>
+        val dataBufSize = kudoTables.map(table => {
+          table.getHeader.getTotalDataLen + table.getHeader.getSerializedSize
+        }).sum
+        val offsetsBufSize = 8 * (kudoTables.length + 1)
+        withResource(Seq(HostMemoryBuffer.allocate(dataBufSize),
+          HostMemoryBuffer.allocate(offsetsBufSize))) { seq =>
+          val dataHost = seq(0)
+          val offsetsHost = seq(1)
+          var currentOffset = 0
+          var i = 0
+          kudoTables.foreach({table =>
```

Comment:
```suggestion
          kudoTables.foreach { table =>
```

but we also need to fix line 371